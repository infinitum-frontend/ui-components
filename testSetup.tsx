import { expect, beforeAll } from 'vitest'
import { render, RenderOptions } from '@testing-library/react'
// import '@testing-library/jest-dom/vitest' for vitest 1.x.x
import { ReactElement, JSXElementConstructor } from 'react'
import matchers from '@testing-library/jest-dom/matchers'

// extends Vitest's expect method with methods from react-testing-library
// https://github.com/testing-library/jest-dom#custom-matchers
expect.extend(matchers)

export const renderComponent = function (
  Component: ReactElement,
  options?: RenderOptions
): {
  el: HTMLElement
  container: HTMLElement
  rerender: (ui: ReactElement<any, string | JSXElementConstructor<any>>) => void
} {
  const { container, rerender } = render(Component, options)

  const el = container.firstChild as HTMLElement

  return { el, container, rerender }
}

beforeAll(() => {
  global.ResizeObserver = class ResizeObserver {
    observe(): any {}
    unobserve(): any {}
    disconnect(): any {}
  }

  // https://github.com/testing-library/react-testing-library/issues/95#issuecomment-392879866
  // https://github.com/vuejs/vue-test-utils/issues/319
  if (!HTMLElement.prototype.scrollTo) {
    HTMLElement.prototype.scrollTo = () => {}
  }
})

// runs a cleanup after each test case (e.g. clearing jsdom)
// afterEach(() => {
//   cleanup()
// })
