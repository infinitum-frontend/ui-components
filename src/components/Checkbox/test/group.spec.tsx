import { describe, it, vi } from 'vitest'
import { renderComponent } from '../../../../testSetup'
import { Checkbox, CheckboxGroup } from '../index'
import userEvent from '@testing-library/user-event'

describe('group', () => {
  const user = userEvent.setup()
  it('should match snapshot', function () {
    const { el } = renderComponent(
      <CheckboxGroup>
        <Checkbox>Label 1</Checkbox>
        <Checkbox>Label 2</Checkbox>
      </CheckboxGroup>
    )

    expect(el).toMatchSnapshot()
  })

  it('should support className', () => {
    const { el } = renderComponent(
      <CheckboxGroup className={'custom-class'}>
        <Checkbox>Label 1</Checkbox>
        <Checkbox>Label 2</Checkbox>
      </CheckboxGroup>
    )
    expect(el).toHaveClass('custom-class')
    expect(el).toHaveClass('inf-checkbox-group')
  })

  it('should be vertical by default', () => {
    const { el } = renderComponent(
      <CheckboxGroup>
        <Checkbox>Label 1</Checkbox>
        <Checkbox>Label 2</Checkbox>
      </CheckboxGroup>
    )

    expect(el).toHaveClass('inf-checkbox-group--direction-vertical')
  })

  it('should support value', () => {
    const { el } = renderComponent(
      <CheckboxGroup value={['1']}>
        <Checkbox value={'0'}>Label 1</Checkbox>
        <Checkbox value={'1'}>Label 2</Checkbox>
      </CheckboxGroup>
    )

    const checkboxList = el.querySelectorAll<HTMLInputElement>(
      'input[type="checkbox"]'
    )
    const checkbox = Array.from(checkboxList).find(
      (item) => item?.value === '1'
    )
    expect(checkbox).toHaveProperty('checked', true)
  })

  it('should call onChange with proper value', async () => {
    let checkedList
    const onChange = vi.fn((value) => (checkedList = value))
    const { el, rerender } = renderComponent(
      <CheckboxGroup onChange={onChange}>
        <Checkbox value={'0'}>Label 1</Checkbox>
        <Checkbox value={'1'}>Label 2</Checkbox>
      </CheckboxGroup>
    )
    await user.click(el.firstChild)
    expect(onChange).toBeCalledTimes(1)
    expect(onChange).toBeCalledWith(checkedList)

    rerender(
      <CheckboxGroup onChange={onChange} value={['0']}>
        <Checkbox value={'0'}>Label 1</Checkbox>
        <Checkbox value={'1'}>Label 2</Checkbox>
      </CheckboxGroup>
    )
    await user.click(el.firstChild)
    expect(onChange).toBeCalledTimes(2)
    expect(checkedList.length).toBe(0)
    expect(onChange).toBeCalledWith(checkedList)
  })
})
