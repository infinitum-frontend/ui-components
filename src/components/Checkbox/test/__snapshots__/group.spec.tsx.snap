// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`group > should match snapshot 1`] = `
<div
  class="inf-checkbox-group inf-checkbox-group--direction-vertical"
>
  <label
    class="inf-checkbox"
  >
    <input
      aria-checked="false"
      aria-required="false"
      type="checkbox"
      value=""
    />
    <span
      class="inf-checkbox__box"
    >
      <svg
        fill="none"
        height="16px"
        viewBox="0 0 20 20"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 9.70588L9.11111 13L13.5 6"
          stroke="currentColor"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </span>
    <span
      class="inf-checkbox__label"
    >
      Label 1
    </span>
  </label>
  <label
    class="inf-checkbox"
  >
    <input
      aria-checked="false"
      aria-required="false"
      type="checkbox"
      value=""
    />
    <span
      class="inf-checkbox__box"
    >
      <svg
        fill="none"
        height="16px"
        viewBox="0 0 20 20"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 9.70588L9.11111 13L13.5 6"
          stroke="currentColor"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </span>
    <span
      class="inf-checkbox__label"
    >
      Label 2
    </span>
  </label>
</div>
`;
